### PR TITLE
[Vice] support for alternative embed-player fixes #13019

### LINKED ISSUE #13019: Cannot download Vice Slutever
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like that [x])
- Use *Preview* tab to see how your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2017.05.07*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2017.05.07**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

C:\>youtube-dl --version
2017.05.07

C:\>youtube-dl -v https://www.vice.com/en_us/article/karley-sciortino-slutever-reloaded
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://www.vice.com/en_us/article/karley-sciortino-slutever-reloaded']
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2017.05.07
[debug] Python version 3.4.4 - Windows-7-6.1.7601-SP1
[debug] exe versions: ffmpeg N-77197-gdf2ce13, ffprobe N-77197-gdf2ce13
[debug] Proxy map: {}
[vice:article] karley-sciortino-slutever-reloaded: Downloading webpage
Traceback (most recent call last):
  File "__main__.py", line 19, in <module>
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\__init__.py", line 465, in main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\__init__.py", line 455, in _real_main
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\YoutubeDL.py", line 1896, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\YoutubeDL.py", line 760, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\extractor\common.py", line 433, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\extractor\vice.py", line 272, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\extractor\common.py", line 791, in _html_search_regex
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpt73_83v_\build\youtube_dl\extractor\common.py", line 761, in _search_regex
  File "C:\Python\Python34\lib\re.py", line 170, in search
TypeError: expected string or buffer

C:\>

### PR DESCRIPTION
**Pull request by Argn0**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

fixes issue  #13019 and several other broken links

### CODE REVIEW

File: youtube_dl/extractor/vice.py [hunk 1]
```diff
@@ -93,7 +104,7 @@ def _extract_preplay_video(self, url, locale, webpage):
 
 class ViceIE(ViceBaseIE):
     IE_NAME = 'vice'
-    _VALID_URL = r'https?://(?:.+?\.)?vice\.com/(?:(?P<locale>[^/]+)/)?videos?/(?P<id>[^/?#&]+)'
+    _VALID_URL = r'https?://(?:.+?\.)?vice\.com/(?:(?P<locale>[^/]+)/)?(?:videos?|embed)/(?P<id>[^/?#&]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/vice.py]: Add a test.

File: youtube_dl/extractor/vice.py [hunk 2]
```diff
@@ -21,10 +21,21 @@
...
-        watch_hub_data = extract_attributes(self._search_regex(
-            r'(?s)(<watch-hub\s*.+?</watch-hub>)', webpage, 'watch hub'))
-        video_id = watch_hub_data['vms-id']
-        title = watch_hub_data['video-title']
+        try:
+            watch_hub_data = extract_attributes(self._search_regex(
+                r'(?s)(<watch-hub\s*.+?</watch-hub>)', webpage, 'watch hub'))
+            video_id = watch_hub_data['vms-id']
+            title = watch_hub_data['video-title']
+        except:
```
Comment (dstftw) [inline youtube_dl/extractor/vice.py]: Never use bare except.

File: youtube_dl/extractor/vice.py [hunk 3]
```diff
@@ -267,8 +281,13 @@ def _url_res(video_url, ie_key):
...
             return _url_res(youtube_url, 'Youtube')
 
-        video_url = self._html_search_regex(
-            r'data-video-url="([^"]+)"',
-            prefetch_data['embed_code'], 'video URL')
+        try:
+            video_url = self._html_search_regex(
+                r'data-video-url="([^"]+)"',
+                prefetch_data['embed_code'], 'video URL')
+        except TypeError:
```
Comment (dstftw) [inline youtube_dl/extractor/vice.py]: Check for `embed_code` instead.

**Comment by Argn0** [comment]
they made changes to the website and I think 'watch_hub_data' is not being used anymore but I left it in just in case

File: youtube_dl/extractor/vice.py [hunk 4]
```diff
@@ -21,13 +21,28 @@
...
+            prefetch_data = prefetch_data.get('video')
+
+            video_id = prefetch_data.get('id')
+            title = prefetch_data.get('title')
+            is_locked = prefetch_data.get('locked') == '1' or prefetch_data.get('locked') == 'true'
+            watch_hub_data = {}
+        else:
+            prefetch_data = {}
+            watch_hub_data = extract_attributes(self._search_regex(
+                r'(?s)(<watch-hub\s*.+?</watch-hub>)', webpage, 'watch hub'))
```
Comment (dstftw) [inline youtube_dl/extractor/vice.py]: Do not change the order of extraction scenarios.

File: youtube_dl/extractor/vice.py [hunk 5]
```diff
@@ -267,8 +288,13 @@ def _url_res(video_url, ie_key):
...
-        video_url = self._html_search_regex(
-            r'data-video-url="([^"]+)"',
-            prefetch_data['embed_code'], 'video URL')
+        if prefetch_data.get('embed_code'):
+            video_url = self._html_search_regex(
+                r'data-video-url="([^"]+)"',
+                prefetch_data['embed_code'], 'video URL')
+        else:
+            video_url = self._html_search_regex(
+                r'<\s*iframe\s*src=\s*"([^"]+)',
```
Comment (dstftw) [inline youtube_dl/extractor/vice.py]: `iframe\s*src` invalid.

File: youtube_dl/extractor/vice.py [hunk 6]
```diff
@@ -21,13 +21,28 @@
...
     def _extract_preplay_video(self, url, locale, webpage):
-        watch_hub_data = extract_attributes(self._search_regex(
-            r'(?s)(<watch-hub\s*.+?</watch-hub>)', webpage, 'watch hub'))
-        video_id = watch_hub_data['vms-id']
-        title = watch_hub_data['video-title']
+
+        prefetch_data = self._parse_json(self._search_regex(
+            r'window\.__PREFETCH_DATA\s*=\s*({.*});',
+            webpage, 'prefetch data'), None, fatal=False)
+        if prefetch_data:
```
Comment (dstftw) [inline youtube_dl/extractor/vice.py]: Add a complete test for this scenario.

**Comment by timofonic** [comment]
@Argn0  @dstftw  Any news?

**Comment by ahmedtds** [comment]
when this will be fixed? vice extractor is broken for a long time!

**Pull request closed without merging**